### ISSUE TITLE
Discussion: optional RankSEG-style decoding support for nnU-Net inference

### ISSUE THREAD

**Issue by Leev1s** [MIC-DKFZ/nnUNet#3050]
Hi nnU-Net maintainers,

I would like to ask whether an optional RankSEG-style decoder could be worth discussing for the nnU-Net inference/export stage. I may be missing nnU-Net-specific constraints, so this is only a discussion issue, not a concrete API request.

I ran a small smoke test using RankSEG on top of an official TotalSegmentator checkpoint in nnU-Net v2 format. The result is not a benchmark, but it looked encouraging enough to ask whether this direction is worth discussing.

## Small Smoke Test

I tested this externally with an official TotalSegmentator organs checkpoint in nnU-Net v2 format:

- checkpoint: `Dataset291_TotalSegmentator_part1_organs_1559subj.zip`
  - https://github.com/wasserth/TotalSegmentator/releases/download/v2.0.0-weights/Dataset291_TotalSegmentator_part1_organs_1559subj.zip
- dataset: TotalSegmentator organs official-test cases
  - https://huggingface.co/datasets/Angelou0516/totalsegmentator-organs
- subset: 20 smallest official-test cases by remote `ct.nii.gz + organs_label.nii.gz` file size
- no retraining or checkpoint modification
- baseline: nnU-Net probability maps decoded with voxel-wise argmax
- RankSEG: `metric="dice"`, `solver="RMA"`, `output_mode="multiclass"`

The metric code is standalone. For this non-region, no-ignore-label demo, it uses the same basic foreground-label averaging convention as nnU-Net-style evaluation: compute per case/label Dice and IoU, average each label across cases with `np.nanmean`, then average foreground label means.

| Method                 | Foreground mean Dice | Foreground mean IoU |
| ---------------------- | -------------------: | ------------------: |
| nnU-Net argmax         |               81.03% |              75.23% |
| RankSEG-style decoding |               81.25% |              75.40% |
| Delta                  |               +0.22% |              +0.17% |

This is not a benchmark. The gain is small, the subset is size-selected, and some labels declined, likely due to the small sample size. The point is only that the method can be inserted after nnU-Net probability prediction without retraining, and there is enough signal to ask whether the direction is reasonable.

Example visualizations:

![Top slice-level Dice gains](https://files.seeusercontent.com/2026/07/02/fh3I/top_slice_gains.png)

## RankSEG References

- RankSEG: A Consistent Ranking-based Framework for Segmentation, Dai and Li, JMLR 2023: https://www.jmlr.org/papers/v24/22-0712.html
- RankSEG-RMA: An Efficient Segmentation Algorithm via Reciprocal Moment Approximation, Wang and Dai: https://arxiv.org/abs/2510.15362
- RankSEG package/source: https://pypi.org/project/rankseg/ and https://github.com/rankseg/rankseg

cc @ZixunWang and @statmlben, with thanks for making the RankSEG work and implementation available.

## Why this may fit nnU-Net

nnU-Net treats segmentation quality as an end-to-end, dataset-dependent pipeline problem: preprocessing, architecture, training, inference, ensembling, and postprocessing are evaluated choices rather than universally fixed assumptions.

From that perspective, voxel-wise argmax is a strong and efficient default, but it is still one probability-to-label decision rule. RankSEG-style decoding could be treated as an optional inference-time candidate.

## Possible Integration Point

If nnU-Net maintainers think this direction is worth exploring, one possible integration point could be the stage where nnU-Net converts model outputs into the final hard segmentation. It would not affect nnU-Net planning, preprocessing, training, checkpoints, or architectures.

From reading the current code, the relevant stage may be around `convert_predicted_logits_to_segmentation_with_correct_shape` in `nnunetv2/inference/export_prediction.py`, where model outputs are converted into the final segmentation. But I am not assuming this is the right API boundary.

```text
nnunetv2/inference/export_prediction.py
convert_predicted_logits_to_segmentation_with_correct_shape(...)
```
## Closing Note

I would mainly appreciate feedback on whether an optional metric-aware decoding mode is compatible with nnU-Net's design goals.

If the maintainers think there is a path forward, we would be happy to help with additional tests, API revisions, documentation, examples, or any engineering work needed to make the contribution easier to review and maintain.

**Comment by wangzixun** [comment]
这是来自qq邮箱的逊已经收到你的邮件。

**Comment by FabianIsensee** [comment]
Thanks for this suggestion! This seems like a very experimental thing to do, rather than an already established and well-proven feature. I am happy to consider adding this if you can provide extensive benchmarks across >10 diverse datasets showing consistent benefits. The last thing I want is a 'hit or miss' type of situation where users may get unexpected drops in segmentation performance on some datasets

**Comment by statmlben** [comment]
Hi @FabianIsensee , thank you for setting a clear benchmark requirement.

I followed up with a fixed, publicly reproducible benchmark across **16 datasets and 2,181 labeled cases**:
- **Release:** https://github.com/rankseg/rankseg-nnunet-benchmark/releases/tag/full16-evidence-v1
- **Full results:** https://github.com/rankseg/rankseg-nnunet-benchmark/blob/full16-evidence-v1/evidence/full16/RESULTS.md
- **Benchmark protocol:** https://github.com/rankseg/rankseg-nnunet-benchmark/blob/full16-evidence-v1/docs/BENCHMARK_PROTOCOL.md

The datasets span CT, MRI, cardiac cine MRI, tumour and organ segmentation, vessels, nested BraTS evaluation regions, and microscopy. Most evaluations use official nnU-Net v1 model releases with out-of-fold predictions; one additional external evaluation exercises the nnU-Net v2 inference path.

For each comparison, RankSEG and argmax receive the same exported model output. Training, checkpoints, preprocessing, evaluation data, and official postprocessing are unchanged. No dataset-specific RankSEG tuning or calibration was fitted on the evaluation cases, and all eligible results—including regressions—were retained.

### Main results
- **13/16 datasets improved; 3/16 regressed**
- Macro-average Dice: **83.63646% argmax** vs **83.84318% RankSEG**
- Dataset-level mean change: **+0.20671 Dice pp**
- Dataset-level median change: **+0.00979 Dice pp**
- At a ±1 Dice pp per-case threshold: **246 improved** | **1,802 stable** | **133 worsened**

As in our original proposal, my intention is only to explore RankSEG as an **explicitly optional inference decoder**. I am not proposing that it replace `argmax` or change nnU-Net's default behavior.

### Calibration follow-up

One limitation is that most official nnU-Net v1 checkpoints used here were trained with the default cross-entropy plus soft-Dice objective. Their exported arrays are softmax-normalized, but normalization alone does not establish probability calibration. Since RankSEG treats its inputs as probabilities, calibration may affect the decoder comparison.

I have opened a matched **CE-only vs CE + soft-Dice calibration study** here:

  https://github.com/rankseg/rankseg-nnunet-benchmark/issues/1

For that study, I can also publicly share the resulting **CE-only checkpoints** and evaluation artifacts so that the comparison can be independently inspected and reproduced.

I would appreciate your view on whether this evidence is useful enough to continue exploring an opt-in integration. I completely understand if you would prefer to see the calibration study, broader nnU-Net v2 evidence, or other validation before considering any code changes.

If you think a small prototype or PR would be helpful, I would also be grateful for guidance on the preferred integration boundary and dependency model.

**Comment by wangzixun** [comment]
这是来自qq邮箱的逊已经收到你的邮件。
